File: packages/contracts/src/bbs-engine.ts
/**
 * BBS engine — element → bar schedule. Pure functions built on the verified
 * IS-code constants in bbs.ts (IS 456 / IS 2502). Given a member's geometry and
 * bar config, it computes each bar's count, cut length and weight, and rolls up
 * the schedule. This is the "enter geometry, get the schedule" automation core;
 * the Estimate app calls it, AORMS only displays the result.
 *
 * Detailing assumptions (all standard, all parameterised with sane defaults):
 *  - straight bar: cut = clearDim − 2·cover + 2·(hook 9d) − 2·(90° bend, 2d)
 *  - stirrup/tie (2-legged, 135° hooks): cut = 2(a+b) + 2·(10d) − 3·(90° bend, 2d)
 *  - column vertical: cut = height + tension lap (one splice per storey)
 */
import {
  HOOK_ALLOWANCE_D,
  BEND_DEDUCTION_D,
  barCount,
  barUnitWeightKgM,
  effectiveCoverMm,
  lapLengthMm,
  type SteelGrade,
} from "./bbs.js";

export type Exposure = "mild" | "moderate" | "severe" | "verySevere" | "extreme";
export type BarRole =
  | "main"
  | "distribution"
  | "top"
  | "stirrup"
  | "tie"
  | "vertical"
  | "bent-up"
  | "dowel";

export interface BbsBar {
  mark: string;
  role: BarRole;
  diaMm: number;
  nos: number;
  cutLengthMm: number;
  unitWtKgM: number;
  weightKg: number;
  shape: string;
}

export interface BbsMember {
  element: "SLAB" | "BEAM" | "COLUMN" | "FOOTING";
  ref?: string;
  bars: BbsBar[];
  totalWeightKg: number;
  byDiameter: { diaMm: number; weightKg: number }[];
}

interface CommonInput {
  ref?: string;
  coverMm?: number;
  exposure?: Exposure;
  concreteGradeMpa: number;
  steelGrade: SteelGrade;
  /** Standard end hook, in bar diameters (default 9d, IS 2502 mild-steel value). */
  hookD?: number;
}

const round = (n: number) => Math.round(n * 100) / 100;

function bar(mark: string, role: BarRole, diaMm: number, nos: number, cutLengthMm: number, shape: string): BbsBar {
  const unitWtKgM = barUnitWeightKgM(diaMm);
  return {
    mark,
    role,
    diaMm,
    nos,
    cutLengthMm: Math.round(cutLengthMm),
    unitWtKgM: round(unitWtKgM),
    weightKg: round((cutLengthMm / 1000) * unitWtKgM * nos),
    shape,
  };
}

function assemble(element: BbsMember["element"], ref: string | undefined, bars: BbsBar[]): BbsMember {
  const byDia = new Map<number, number>();
  let total = 0;
  for (const b of bars) {
    total += b.weightKg;
    byDia.set(b.diaMm, (byDia.get(b.diaMm) ?? 0) + b.weightKg);
  }
  return {
    element,
    ref,
    bars,
    totalWeightKg: round(total),
    byDiameter: [...byDia.entries()]
      .sort((a, b) => a[0] - b[0])
      .map(([diaMm, weightKg]) => ({ diaMm, weightKg: round(weightKg) })),
  };
}

/** Straight bar cut length: clearDim − 2·cover + 2·hook − 2·(90° bend deduction). */
function straightCut(clearDimMm: number, coverMm: number, diaMm: number, hookD: number): number {
  return clearDimMm - 2 * coverMm + 2 * hookD * diaMm - 2 * BEND_DEDUCTION_D[90] * diaMm;
}

/** 2-legged closed stirrup/tie cut length around an (a × b) core with 135° hooks. */
function stirrupCut(sideAMm: number, sideBMm: number, diaMm: number): number {
  const a = sideAMm; // core width  = section − 2·cover
  const b = sideBMm; // core depth  = section − 2·cover
  const hook = 10 * diaMm; // 135° hook, standard 10d each
  return 2 * (a + b) + 2 * hook - 3 * BEND_DEDUCTION_D[90] * diaMm;
}

// ── Slab (one-way: bottom main + distribution) ────────────────────────────────
export interface SlabInput extends CommonInput {
  lengthMm: number; // span in the main-bar direction
  widthMm: number;
  mainDiaMm: number;
  mainSpacingMm: number;
  distDiaMm: number;
  distSpacingMm: number;
}
export function computeSlabBBS(i: SlabInput): BbsMember {
  const cover = i.coverMm ?? effectiveCoverMm("slab", i.exposure ?? "mild");
  const hookD = i.hookD ?? HOOK_ALLOWANCE_D[2]; // 9d
  const main = bar(
    "S1",
    "main",
    i.mainDiaMm,
    barCount(i.widthMm, i.mainSpacingMm),
    straightCut(i.lengthMm, cover, i.mainDiaMm, hookD),
    "straight",
  );
  const dist = bar(
    "S2",
    "distribution",
    i.distDiaMm,
    barCount(i.lengthMm, i.distSpacingMm),
    straightCut(i.widthMm, cover, i.distDiaMm, hookD),
    "straight",
  );
  return assemble("SLAB", i.ref, [main, dist]);
}

// ── Beam (simply supported: bottom + top longitudinal + 2-legged stirrups) ─────
export interface BeamInput extends CommonInput {
  clearSpanMm: number;
  widthMm: number;
  depthMm: number;
  bottomDiaMm: number;
  bottomNos: number;
  topDiaMm: number;
  topNos: number;
  stirrupDiaMm: number;
  stirrupSpacingMm: number;
}
export function computeBeamBBS(i: BeamInput): BbsMember {
  const cover = i.coverMm ?? effectiveCoverMm("beam", i.exposure ?? "moderate");
  const hookD = i.hookD ?? HOOK_ALLOWANCE_D[2];
  const bottom = bar(
    "B1",
    "main",
    i.bottomDiaMm,
    i.bottomNos,
    straightCut(i.clearSpanMm, cover, i.bottomDiaMm, hookD),
    "straight",
  );
  const top = bar("B2", "top", i.topDiaMm, i.topNos, straightCut(i.clearSpanMm, cover, i.topDiaMm, hookD), "straight");
  const stirrup = bar(
    "B3",
    "stirrup",
    i.stirrupDiaMm,
    barCount(i.clearSpanMm, i.stirrupSpacingMm),
    stirrupCut(i.widthMm - 2 * cover, i.depthMm - 2 * cover, i.stirrupDiaMm),
    "stirrup",
  );
  return assemble("BEAM", i.ref, [bottom, top, stirrup]);
}

// ── Column (vertical bars + lateral ties) ─────────────────────────────────────
export interface ColumnInput extends CommonInput {
  heightMm: number; // storey clear height
  widthMm: number;
  depthMm: number;
  verticalDiaMm: number;
  verticalNos: number;
  tieDiaMm: number;
  tieSpacingMm: number;
}
export function computeColumnBBS(i: ColumnInput): BbsMember {
  const cover = i.coverMm ?? effectiveCoverMm("column", i.exposure ?? "moderate");
  const lap = lapLengthMm(i.verticalDiaMm, i.steelGrade, i.concreteGradeMpa, "compression");
  const vertical = bar("C1", "vertical", i.verticalDiaMm, i.verticalNos, i.heightMm + lap, "straight+lap");
  const tie = bar(
    "C2",
    "tie",
    i.tieDiaMm,
    barCount(i.heightMm, i.tieSpacingMm),
    stirrupCut(i.widthMm - 2 * cover, i.depthMm - 2 * cover, i.tieDiaMm),
    "tie",
  );
  return assemble("COLUMN", i.ref, [vertical, tie]);
}

// ── Footing (two-way bottom mesh) ─────────────────────────────────────────────
export interface FootingInput extends CommonInput {
  lengthMm: number;
  widthMm: number;
  xDiaMm: number;
  xSpacingMm: number;
  yDiaMm: number;
  ySpacingMm: number;
}
export function computeFootingBBS(i: FootingInput): BbsMember {
  const cover = i.coverMm ?? effectiveCoverMm("footing", i.exposure ?? "moderate");
  const hookD = i.hookD ?? HOOK_ALLOWANCE_D[2];
  const xbar = bar(
    "F1",
    "main",
    i.xDiaMm,
    barCount(i.widthMm, i.xSpacingMm),
    straightCut(i.lengthMm, cover, i.xDiaMm, hookD),
    "straight",
  );
  const ybar = bar(
    "F2",
    "distribution",
    i.yDiaMm,
    barCount(i.lengthMm, i.ySpacingMm),
    straightCut(i.widthMm, cover, i.yDiaMm, hookD),
    "straight",
  );
  return assemble("FOOTING", i.ref, [xbar, ybar]);
}

export type MemberInput =
  | ({ element: "SLAB" } & SlabInput)
  | ({ element: "BEAM" } & BeamInput)
  | ({ element: "COLUMN" } & ColumnInput)
  | ({ element: "FOOTING" } & FootingInput);

/** Dispatch to the right element engine. */
export function computeMemberBBS(input: MemberInput): BbsMember {
  switch (input.element) {
    case "SLAB":
      return computeSlabBBS(input);
    case "BEAM":
      return computeBeamBBS(input);
    case "COLUMN":
      return computeColumnBBS(input);
    case "FOOTING":
      return computeFootingBBS(input);
  }
}

/** Roll up many members into one schedule grouped by diameter. */
export function scheduleByDiameter(members: BbsMember[]): { diaMm: number; weightKg: number }[] {
  const byDia = new Map<number, number>();
  for (const m of members)
    for (const d of m.byDiameter) byDia.set(d.diaMm, (byDia.get(d.diaMm) ?? 0) + d.weightKg);
  return [...byDia.entries()].sort((a, b) => a[0] - b[0]).map(([diaMm, weightKg]) => ({ diaMm, weightKg: round(weightKg) }));
}

File: estimate/backend-cpp/src/bbs_engine.cpp
#include "bbs_engine.hpp"

#include <algorithm>
#include <cmath>
#include <map>
#include <stdexcept>

#include "bbs.hpp"
#include "numeric.hpp"

namespace aorms {
namespace {

// ── JSON field accessors (mirror the TS `?? default` semantics) ──────────────
double reqNum(const json& j, const char* key) {
  if (!j.contains(key) || j.at(key).is_null() || !j.at(key).is_number())
    throw std::invalid_argument(std::string("BBS member missing numeric field: ") + key);
  return j.at(key).get<double>();
}

double optNum(const json& j, const char* key, double fallback) {
  auto it = j.find(key);
  if (it == j.end() || it->is_null() || !it->is_number()) return fallback;
  return it->get<double>();
}

std::string optStr(const json& j, const char* key, const std::string& fallback) {
  auto it = j.find(key);
  if (it == j.end() || it->is_null() || !it->is_string()) return fallback;
  return it->get<std::string>();
}

bool has(const json& j, const char* key) {
  auto it = j.find(key);
  return it != j.end() && !it->is_null();
}

// bar(): unit weight from D²/162; cut length rounded to a whole mm for display,
// but the weight is computed from the UNROUNDED cut length and unit weight —
// exactly as bbs-engine.ts does.
BbsBar bar(const std::string& mark, const std::string& role, double diaMm, int nos,
           double cutLengthMm, const std::string& shape) {
  double unitWtKgM = bbs::barUnitWeightKgM(diaMm);
  BbsBar b;
  b.mark = mark;
  b.role = role;
  b.diaMm = diaMm;
  b.nos = nos;
  b.cutLengthMm = static_cast<int>(jsRound(cutLengthMm));
  b.unitWtKgM = round2(unitWtKgM);
  b.weightKg = round2((cutLengthMm / 1000.0) * unitWtKgM * nos);
  b.shape = shape;
  return b;
}

BbsMember assemble(const std::string& element, const std::string& ref, std::vector<BbsBar> bars) {
  std::map<double, double> byDia; // ordered ascending by diameter
  double total = 0;
  for (const auto& b : bars) {
    total += b.weightKg;
    byDia[b.diaMm] += b.weightKg;
  }
  BbsMember m;
  m.element = element;
  m.ref = ref;
  m.bars = std::move(bars);
  m.totalWeightKg = round2(total);
  for (const auto& [dia, w] : byDia) m.byDiameter.push_back({dia, round2(w)});
  return m;
}

// straight bar cut: clearDim − 2·cover + 2·hook − 2·(90° bend deduction)
double straightCut(double clearDimMm, double coverMm, double diaMm, double hookD) {
  return clearDimMm - 2 * coverMm + 2 * hookD * diaMm - 2 * bbs::bendDeductionD(90) * diaMm;
}

// 2-legged closed stirrup/tie around an (a × b) core, 135° hooks (10d each)
double stirrupCut(double aMm, double bMm, double diaMm) {
  double hook = 10 * diaMm;
  return 2 * (aMm + bMm) + 2 * hook - 3 * bbs::bendDeductionD(90) * diaMm;
}

BbsMember computeSlab(const json& i, const std::string& ref) {
  double cover = has(i, "coverMm") ? reqNum(i, "coverMm")
                                   : bbs::effectiveCoverMm("slab", optStr(i, "exposure", "mild"));
  double hookD = optNum(i, "hookD", bbs::hookAllowanceD(2));
  double lengthMm = reqNum(i, "lengthMm"), widthMm = reqNum(i, "widthMm");
  BbsBar main = bar("S1", "main", reqNum(i, "mainDiaMm"),
                    bbs::barCount(widthMm, reqNum(i, "mainSpacingMm")),
                    straightCut(lengthMm, cover, reqNum(i, "mainDiaMm"), hookD), "straight");
  BbsBar dist = bar("S2", "distribution", reqNum(i, "distDiaMm"),
                    bbs::barCount(lengthMm, reqNum(i, "distSpacingMm")),
                    straightCut(widthMm, cover, reqNum(i, "distDiaMm"), hookD), "straight");
  return assemble("SLAB", ref, {main, dist});
}

BbsMember computeBeam(const json& i, const std::string& ref) {
  double cover = has(i, "coverMm") ? reqNum(i, "coverMm")
                                   : bbs::effectiveCoverMm("beam", optStr(i, "exposure", "moderate"));
  double hookD = optNum(i, "hookD", bbs::hookAllowanceD(2));
  double clearSpanMm = reqNum(i, "clearSpanMm");
  double widthMm = reqNum(i, "widthMm"), depthMm = reqNum(i, "depthMm");
  BbsBar bottom = bar("B1", "main", reqNum(i, "bottomDiaMm"), static_cast<int>(reqNum(i, "bottomNos")),
                      straightCut(clearSpanMm, cover, reqNum(i, "bottomDiaMm"), hookD), "straight");
  BbsBar top = bar("B2", "top", reqNum(i, "topDiaMm"), static_cast<int>(reqNum(i, "topNos")),
                   straightCut(clearSpanMm, cover, reqNum(i, "topDiaMm"), hookD), "straight");
  BbsBar stirrup = bar("B3", "stirrup", reqNum(i, "stirrupDiaMm"),
                       bbs::barCount(clearSpanMm, reqNum(i, "stirrupSpacingMm")),
                       stirrupCut(widthMm - 2 * cover, depthMm - 2 * cover, reqNum(i, "stirrupDiaMm")),
                       "stirrup");
  return assemble("BEAM", ref, {bottom, top, stirrup});
}

BbsMember computeColumn(const json& i, const std::string& ref) {
  double cover = has(i, "coverMm") ? reqNum(i, "coverMm")
                                   : bbs::effectiveCoverMm("column", optStr(i, "exposure", "moderate"));
  double vDia = reqNum(i, "verticalDiaMm");
  double lap = bbs::lapLengthMm(vDia, optStr(i, "steelGrade", "Fe500"), reqNum(i, "concreteGradeMpa"), true);
  double heightMm = reqNum(i, "heightMm");
  double widthMm = reqNum(i, "widthMm"), depthMm = reqNum(i, "depthMm");
  BbsBar vertical = bar("C1", "vertical", vDia, static_cast<int>(reqNum(i, "verticalNos")),
                        heightMm + lap, "straight+lap");
  BbsBar tie = bar("C2", "tie", reqNum(i, "tieDiaMm"),
                   bbs::barCount(heightMm, reqNum(i, "tieSpacingMm")),
                   stirrupCut(widthMm - 2 * cover, depthMm - 2 * cover, reqNum(i, "tieDiaMm")), "tie");
  return assemble("COLUMN", ref, {vertical, tie});
}

BbsMember computeFooting(const json& i, const std::string& ref) {
  double cover = has(i, "coverMm") ? reqNum(i, "coverMm")
                                   : bbs::effectiveCoverMm("footing", optStr(i, "exposure", "moderate"));
  double hookD = optNum(i, "hookD", bbs::hookAllowanceD(2));
  double lengthMm = reqNum(i, "lengthMm"), widthMm = reqNum(i, "widthMm");
  BbsBar xbar = bar("F1", "main", reqNum(i, "xDiaMm"),
                    bbs::barCount(widthMm, reqNum(i, "xSpacingMm")),
                    straightCut(lengthMm, cover, reqNum(i, "xDiaMm"), hookD), "straight");
  BbsBar ybar = bar("F2", "distribution", reqNum(i, "yDiaMm"),
                    bbs::barCount(lengthMm, reqNum(i, "ySpacingMm")),
                    straightCut(widthMm, cover, reqNum(i, "yDiaMm"), hookD), "straight");
  return assemble("FOOTING", ref, {xbar, ybar});
}

} // namespace

BbsMember computeMemberBBS(const json& input) {
  std::string element = optStr(input, "element", "");
  std::string ref = optStr(input, "ref", "");
  if (element == "SLAB") return computeSlab(input, ref);
  if (element == "BEAM") return computeBeam(input, ref);
  if (element == "COLUMN") return computeColumn(input, ref);
  if (element == "FOOTING") return computeFooting(input, ref);
  throw std::invalid_argument("unknown BBS element: " + element);
}

std::vector<ByDiameter> scheduleByDiameter(const std::vector<BbsMember>& members) {
  std::map<double, double> byDia;
  for (const auto& m : members)
    for (const auto& d : m.byDiameter) byDia[d.diaMm] += d.weightKg;
  std::vector<ByDiameter> out;
  for (const auto& [dia, w] : byDia) out.push_back({dia, round2(w)});
  return out;
}

} // namespace aorms
